src/particles/elements/RFCavity.H

```diff
@@ -0,0 +1,391 @@
+            amrex::ParticleReal const s = refpart.s;
+            amrex::ParticleReal const sedge = refpart.sedge;
+
+            // initialize linear map (deviation) values
+            for (int i=1; i<7; i++) {
+               for (int j=1; j<7; j++) {
+                  if (i == j)
+                      refpart.map(i,j) = 1.0_prt;
+                  else
+                      refpart.map(i,j) = 0.0_prt;
+               }
+            }
+
+            // length of the current slice
+            amrex::ParticleReal const slice_ds = m_ds / nslice();
+
+            // compute intial value of beta*gamma
+            amrex::ParticleReal const bgi = sqrt(pow(pt,2)-1.0_prt);
+
+            // call integrator to advance (t,pt)
+            amrex::ParticleReal const zin = s - sedge;
+            amrex::ParticleReal const zout = zin + slice_ds;
+            std::string element_type = "rfcavity";
+            int const nsteps = m_mapsteps;
+
+            integrators::symp2_integrate_split3(refpart,zin,zout,nsteps,*this);
+            amrex::ParticleReal const ptf = refpart.pt;
+
+            // advance position (x,y,z)
+            refpart.x = x + slice_ds*px/bgi;
+            refpart.y = y + slice_ds*py/bgi;
+            refpart.z = z + slice_ds*pz/bgi;
+
+            // compute final value of beta*gamma
+            amrex::ParticleReal const bgf = sqrt(pow(ptf,2)-1.0_prt);
+
+            // advance momentum (px,py,pz)
+            refpart.px = px*bgf/bgi;
+            refpart.py = py*bgf/bgi;
+            refpart.pz = pz*bgf/bgi;
+
+            // convert linear map from dynamic to static units
+            amrex::ParticleReal scale_in = 1.0_prt;
+            amrex::ParticleReal scale_fin = 1.0_prt;
+
+            for (int i=1; i<7; i++) {
+               for (int j=1; j<7; j++) {
+                   if( i % 2 == 0)
+                      scale_fin = bgf;
+                   else
+                      scale_fin = 1.0_prt;
+                   if( j % 2 == 0)
+                      scale_in = bgi;
+                   else
+                      scale_in = 1.0_prt;
+                   refpart.map(i,j) = refpart.map(i,j)*scale_in/scale_fin;
+               }
+            }
+
+            // advance integrated path length
+            refpart.s = s + slice_ds;
+        }
+
+        std::tuple<amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal>
+        RF_Efield (amrex::ParticleReal const zeval) const
+        {
+            using namespace amrex::literals; // for _rt and _prt
+
+            // specify constants
+            amrex::ParticleReal const pi = 3.141592653589793_prt;
+            amrex::ParticleReal const zlen = 1.31879807_prt;
+            amrex::ParticleReal const zmid = zlen/2.0_prt;
+            int const ncoef = 25;
+
+            // specify Fourier coefficients
+            amrex::ParticleReal cos_coef [ncoef] = {};
+            amrex::ParticleReal sin_coef [ncoef] = {};
+
+            // recall array indexing begins with 0
```

Comment:
I do not understand where these numerical constants come from.

Reply:
Will add a comment here specifying the origin of these constants.

These are Fourier coefficients specifying the on-axis longitudinal electric field in the default 9-cell TESLA RF cavity.  Although they will remain as default values, they can be superseded by user-provided input in a follow-up PR.

Reply:
Will add a comment + reference here.